Fix overlay-qsv

ruiling opened the pull request.

ruiling-song committed:
lavf: make overlay_qsv work based on framesync.

the previous version which was cherry-picked does not work
with FFmpeg framework, because ff_request_frame() was totally
different between Libav and FFmpeg. So, I make it work through
using framesync. There are still some issues to be fixed in
later patch.

v2:
add .preinit field to initilize framesync options.
export more options like vf_overlay.c

Signed-off-by: Ruiling Song <ruiling.song@intel.com>

ruiling-song committed:
lavf: clone the frame managed by framework.

we should clone the frame, which is managed by the framework.
directly assign it will cause double-free issue when qsv try
to free it. In fact, the frames was managed by the framework!

Right now, I am still not quite sure why we receive 'more data'
error from libmfx. But some simple debugging seems that it
is non-sense. so just skip it totally, not bothering to
return a EAGAIN error to the caller.

Signed-off-by: Ruiling Song <ruiling.song@intel.com>

lizhong1008 commented:
Testing with 
./ffmpeg -hwaccel qsv -c:v h264_qsv -analyzeduration 10000000 -i  /samba/anonymous/Videos/libav_cavlc0.h264 -an -movflags faststart -qmax 51 -qmin 10 -preset faster -level 3.0 -b:v 734537 -c:v h264_qsv -codec:a libfdk_aac -map_metadata 0:g -vf '[0:v]vpp_qsv=w=trunc(in_w/2)*2:h=trunc(in_h/2)*2[scale2],[scale2]vpp_qsv=w=480:h=480[scale1],movie=/samba/anonymous/Videos/ffmpeg_fate_suite_lastest/fate-suite/png1/lena-rgba.png,hwupload,format=qsv[wm];[scale1][wm]overlay_qsv=x_expr=iw(0)-iw(1)-10+0:y_expr=10+0[wmimage]' -f mp4 -y overlay_3.3.1.mp4

Segment fault.
 (gdb) bt
#0  0x00000000006c4115 in submit_frame (s=0x3278420, inlink=0x2687a60, picref=0x0) at libavfilter/qsvvpp.c:295
#1  0x00000000006c5411 in ff_qsvvpp_filter_frame (s=0x3278420, inlink=0x2687a60, picref=0x0) at libavfilter/qsvvpp.c:686
#2  0x0000000000568ef2 in process_frame (fs=0x2d645c8) at libavfilter/vf_overlay_qsv.c:385
#3  0x00000000004750c3 in ff_framesync_activate (fs=0x2d645c8) at libavfilter/framesync.c:353
#4  0x0000000000569620 in activate (ctx=0x2d644c0) at libavfilter/vf_overlay_qsv.c:544
#5  0x000000000045e3c4 in ff_filter_activate (filter=0x2d644c0) at libavfilter/avfilter.c:1500
#6  0x0000000000462cca in ff_filter_graph_run_once (graph=0x29eed60) at libavfilter/avfiltergraph.c:1449
#7  0x0000000000462eb7 in get_frame_internal (ctx=0x2b98de0, frame=0x0, flags=1, samples=0) at libavfilter/buffersink.c:110
#8  0x0000000000462f14 in av_buffersink_get_frame_flags (ctx=0x2b98de0, frame=0x0, flags=1) at libavfilter/buffersink.c:121
#9  0x0000000000462a5d in avfilter_graph_request_oldest (graph=0x29eed60) at libavfilter/avfiltergraph.c:1402
#10 0x0000000000430bfa in transcode_from_filter (graph=0x25e4620, best_ist=0x7fffffffd6b8) at fftools/ffmpeg.c:4473
#11 0x0000000000430f47 in transcode_step () at fftools/ffmpeg.c:4548
#12 0x00000000004311bf in transcode () at fftools/ffmpeg.c:4624
#13 0x00000000004318ab in main (argc=34, argv=0x7fffffffd948) at fftools/ffmpeg.c:4830

It can work well on qsv331 branch.

ruiling-song committed:
lavf/qsv: Support multiple inputs overlay composition

Based on patch by Liu, ChaoX.
    lavfi/vf_overlay_qsv: support mulit-inputs overlay.

    Description:
    1. layout(grid/manual/overlay) default is the preset 'overlay' mode
    2. x_expr,y_expr,w_expr,z_expr,a_expr
        If layout is set to overlay, these exprs will be used to calculate
    the [overlay] channel's output rect while the [main] channle will be
    forced the same as the 1st input.
        If layout is set to manual, these expressions will be used to
    construct the layout. Some const keys and funcs can be used.
        If keyword 'array' is used, the expressions must have enough
    comma-seprated statements to cover each input or it will cause an error.
    Otherwise, the single statement will be applied to each window.

    3. Const values
        @array(\,) x_expr=array(expr1\,expr2\,...\,exprn) an array of
                   expressions indicating each windows x/y/w/h
        @x(n),y(n),w(n),h(n),alpha(n),iw(n),ih(n) a group of functions that can get
        each windows position attributions, where n indicates the index
        counting from 0.
        @N,n Const values, N indicates the number of input windows while n
        indicates the index of current window.

    Examples:
        1. ffmpeg -hwaccel qsv -c:v h264_qsv -i test.mp4 -filter_complex
           "split=4,overlay_qsv=4:manual:x_expr=w(n)*mod(n\,ceil(sqrt(N))):y_expr=h(n)*floor(n/ceil(sqrt(N))):w_expr=W/ceil(sqrt(N)):h_expr=H/ceil(sqrt(N))"
           -preset veryfast -c:v h264_qsv output.mp4
           When calculating the layout, 'n' will be replaced by the index of current window and this is the way how 'grid' mode calculate.
        2. ffmpeg -hwaccel qsv -c:v h264_qsv -i test.mp4 -filter_complex
           "split=4,overlay_qsv=4:manual:x_expr=array(0\,960\,0,960):y_expr=array(0\,0\,540\,540):w_expr=array(960,960,960,960):h_expr=array(540,540,540,540):a_expr=255"
           -preset veryfast -c:v h264_qsv output.mp4
           This command is equal to the former one when test.mp4 is of FHD.
        3. to overlay a video to area [(0,0), (256,256)] on another video, you can use:
           ffmpeg -hwaccel qsv -c:v h264_qsv -i test1.mp4 -hwaccel qsv -c:v h264_qsv -i test2.mp4 -filter_complex
           "overlay_qsv=x_expr=0:y_expr=0:w_expr=256:h_expr=256:a_expr=128" -preset veryfast -c:v h264_qsv output.mp4

Signed-off-by: Ruiling Song <ruiling.song@intel.com>

lizhong1008 commented:
Patches applied.
